.changeset/ten-cougars-pretend.md

```diff
@@ -0,0 +1,84 @@
+---
+'@astrojs/netlify': major
+---
+
+# Netlify Adapter v4 simplifies static + SSR deployments
+
+This update is a complete overhaul of the Netlify adapter.
+It simplifies the user-facing config, and resolves a number of bugs along the way.
+
+Here's what changes:
+
+## Middleware without configuration
+
+In v3, [Middleware](https://docs.astro.build/en/guides/middleware/) was deployed to the SSR Function by default.
+Enabling the `edgeMiddleware` option would change this behaviour to deploy it using Netlify Edge Functions instead, improving
+page speed and enabling the middleware to run on prerendered sites.
+
+In v4, this is the default setting.
+If your Astro site makes use of Middleware,
+it's deployed to an Edge Function where it runs super close to your users and is also applied to static sites.
+You can use this to implement customization or authorization logic!
```

Comment:
I am not sure we should enable this by default. Edge functions are subject to billing, especially in the free tier. If the edge functions happen to use all the bandwidth of the user, they will start paying without even noticing. What do you think?

Reply:
The free tier includes 1M edge function executions (see https://www.netlify.com/pricing/#core-pricing-table), which should be more than enough. Also, if there's no Middleware in the Astro project, then no Edge Function is generated. I think it's fine to have this as a default.

Reply:
The Astro docs say the following:

> Middleware allows you to intercept requests and responses and inject behaviors dynamically every time a page or endpoint is about to be rendered.
> ....
> Middleware is available in both SSG and SSR Astro projects.

To me, this reads like Middleware should be executed not only for on-demand-rendered pages, but also for static pages. The only way of implementing this on Netlify is Edge Functions, as Functions don't run in front of static pages.